node/lease_controller_v1.go

```diff
@@ -0,0 +1,341 @@
+
+	var (
+		lease   *coordinationv1.Lease
+		created bool
+		err     error
+	)
+	sleep := 100 * time.Millisecond
+	for {
+		lease, created, err = c.ensureLease(ctx, node)
+		if err == nil {
+			break
+		}
+		sleep = minDuration(2*sleep, maxBackoff)
+		log.G(ctx).WithError(err).Errorf("failed to ensure node lease exists, will retry in %v", sleep)
+		// backoff wait
+		c.clock.Sleep(sleep)
+		timer := c.clock.NewTimer(sleep)
+		defer timer.Stop()
+		select {
+		case <-timer.C():
+		case <-ctx.Done():
+			return nil, false
+		}
+	}
+	return lease, created
+}
+
+// ensureLease creates the lease if it does not exist. Returns the lease and
+// a bool (true if this call created the lease), or any error that occurs.
+func (c *leaseController) ensureLease(ctx context.Context, node *corev1.Node) (*coordinationv1.Lease, bool, error) {
+	ctx, span := trace.StartSpan(ctx, "controller.ensureLease")
+	defer span.End()
+
+	lease, err := c.leaseClient.Get(ctx, node.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// lease does not exist, create it.
+		leaseToCreate := c.newLease(ctx, node, nil)
+		if len(leaseToCreate.OwnerReferences) == 0 {
+			// We want to ensure that a lease will always have OwnerReferences set.
+			// Thus, given that we weren't able to set it correctly, we simply
+			// not create it this time - we will retry in the next iteration.
+			return nil, false, nil
+		}
+		lease, err := c.leaseClient.Create(ctx, leaseToCreate, metav1.CreateOptions{})
+		if err != nil {
+			span.SetStatus(err)
+			return nil, false, err
+		}
+		log.G(ctx).Debug("Successfully created lease")
+		return lease, true, nil
+	} else if err != nil {
+		// unexpected error getting lease
+		log.G(ctx).WithError(err).Error("Unexpected error getting lease")
+		span.SetStatus(err)
+		return nil, false, err
+	}
+	log.G(ctx).Debug("Successfully recovered existing lease")
+	// lease already existed
+	return lease, false, nil
+}
+
+// retryUpdateLease attempts to update the lease for maxUpdateRetries,
+// call this once you're sure the lease has been created
+func (c *leaseController) retryUpdateLease(ctx context.Context, node *corev1.Node, base *coordinationv1.Lease) error {
+	ctx, span := trace.StartSpan(ctx, "controller.retryUpdateLease")
+	defer span.End()
+
+	for i := 0; i < maxUpdateRetries; i++ {
+		lease, err := c.leaseClient.Update(ctx, c.newLease(ctx, node, base), metav1.UpdateOptions{})
+		if err == nil {
+			log.G(ctx).WithField("retries", i).Debug("Successfully updated lease")
+			c.latestLease = lease
+			return nil
+		}
+		log.G(ctx).WithError(err).Error("failed to update node lease")
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			break
+		}
+		// OptimisticLockError requires getting the newer version of lease to proceed.
+		if apierrors.IsConflict(err) {
+			base, _ = c.backoffEnsureLease(ctx, node)
+			continue
+		}
+	}
+	err := fmt.Errorf("failed %d attempts to update node lease", maxUpdateRetries)
```

Comment:
Not all cases are `maxUpdateRetries`
We are also loose the error from the loop.